src/java.base/share/classes/java/lang/invoke/TypeConvertingMethodAdapter.java

```diff
@@ -204,92 +133,76 @@ private Wrapper toWrapper(String desc) {
      * @param target
      * @param functional
      */
-    void convertType(Class<?> arg, Class<?> target, Class<?> functional) {
+    static void convertType(CodeBuilder cob, Class<?> arg, Class<?> target, Class<?> functional) {
         if (arg.equals(target) && arg.equals(functional)) {
             return;
         }
         if (arg == Void.TYPE || target == Void.TYPE) {
             return;
         }
         if (arg.isPrimitive()) {
-            Wrapper wArg = Wrapper.forPrimitiveType(arg);
             if (target.isPrimitive()) {
                 // Both primitives: widening
-                widen(wArg, Wrapper.forPrimitiveType(target));
+                widen(cob, TypeKind.from(arg), TypeKind.from(target));
             } else {
                 // Primitive argument to reference target
-                String dTarget = BytecodeDescriptor.unparse(target);
-                Wrapper wPrimTarget = wrapperOrNullFromDescriptor(dTarget);
-                if (wPrimTarget != null) {
+                TypeKind wPrimTk = primitiveTypeKindFromClass(target);
+                if (wPrimTk != null) {
                     // The target is a boxed primitive type, widen to get there before boxing
-                    widen(wArg, wPrimTarget);
-                    box(wPrimTarget);
+                    widen(cob, TypeKind.from(arg), wPrimTk);
+                    box(cob, wPrimTk);
                 } else {
                     // Otherwise, box and cast
-                    box(wArg);
-                    cast(wrapperName(wArg), dTarget);
+                    box(cob, TypeKind.from(arg));
+                    cast(cob, classDesc(target));
                 }
             }
         } else {
-            String dArg = BytecodeDescriptor.unparse(arg);
-            String dSrc;
-            if (functional.isPrimitive()) {
-                dSrc = dArg;
+            Class<?> src;
+            if (arg == functional || functional.isPrimitive()) {
```

Comment:
`arg == functional` avoids a cast, though it's not in parity with the older code.